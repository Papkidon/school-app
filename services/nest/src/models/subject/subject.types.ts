import { z } from 'zod';

export const subjectCreateSchema = z.object({
  data: z.object({
    name: z.string(),
    points: z.number(),
  }),
});

export const subjectUpdateSchema = subjectCreateSchema.deepPartial();

export type subjectsWithClassesDto = {
  result: {
    id: string;
    name: string;
    classes: {
      id: string;
      number: string;
      max_population: number;
    }[];
  };
}[];

export type mappedSubjectsWithClassesDto = {
  id: string;
  name: string;
  classes: { id: string; number: string; max_population: number }[];
}[];

export type subjectType = {
  data: {
    name: string;
    points: number;
  };
};
