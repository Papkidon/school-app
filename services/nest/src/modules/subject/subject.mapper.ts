import {
  mappedSubjectsWithClassesDto,
  subjectsWithClassesDto,
} from '../../models/subject/subject.types';

export const mapSubjectsWithClasses = (
  subjects: subjectsWithClassesDto,
): mappedSubjectsWithClassesDto => {
  if (!subjects.length) return [];
  return subjects.map((subject) => {
    return {
      ...subject.result,
    };
  });
};
