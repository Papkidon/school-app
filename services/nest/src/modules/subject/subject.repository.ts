import { Injectable } from '@nestjs/common';
import { Subject } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import {
  subjectsWithClassesDto,
  subjectType,
} from '../../models/subject/subject.types';
import { DeepPartial } from '../../models/utility.types';

@Injectable()
export class SubjectRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<Subject[]> {
    return await this.prisma.subject.findMany();
  }

  async findById(id: string): Promise<Subject> {
    return await this.prisma.subject.findUnique({ where: { id } });
  }

  async findByName(name: string): Promise<Subject> {
    return await this.prisma.subject.findUnique({ where: { name } });
  }

  async findAllSubjectsWithClasses() {
    return (await this.prisma.$queryRaw`
        SELECT
        json_build_object (
            'id', s.id,
            'name', s.name,
            'points', s.points,
            'classes', jsonb_agg (c)
            ) AS result
       FROM "Subject" s
       LEFT JOIN "ClassSubject" as cs ON
       s.id = cs.subject_id
       LEFT JOIN "Class" as c ON
       cs.class_id = c.id
       GROUP BY s.id;`) as subjectsWithClassesDto;
  }

  async create(body: any): Promise<Subject> {
    return await this.prisma.subject.create(body);
  }

  async delete(id: string): Promise<Subject> {
    return await this.prisma.subject.delete({ where: { id } });
  }

  async update(id: string, body: DeepPartial<subjectType>): Promise<Subject> {
    return await this.prisma.subject.update({
      where: { id },
      data: { ...body.data },
    });
  }
}
