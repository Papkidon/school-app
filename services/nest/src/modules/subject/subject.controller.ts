import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Patch,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { SubjectService } from './subject.service';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import {
  subjectCreateSchema,
  subjectType,
  subjectUpdateSchema,
} from '../../models/subject/subject.types';
import { DeepPartial } from '../../models/utility.types';
import { Subject } from '@prisma/client';
import { mapSubjectsWithClasses } from './subject.mapper';
import { idSchema } from '../../models/utility.types';

@Controller('api/v1/subject')
export class SubjectController {
  constructor(private readonly subjectService: SubjectService) {}

  @Get('')
  async findAll(): Promise<Subject[]> {
    try {
      const result = await this.subjectService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No user data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Get('classes')
  async findAllSubjectsWithClasses(): Promise<unknown> {
    try {
      return mapSubjectsWithClasses(
        await this.subjectService.findAllSubjectWithClasses(),
      );
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(subjectCreateSchema))
    createSubjectDto: subjectType,
  ): Promise<Subject> {
    try {
      return await this.subjectService.create(createSubjectDto);
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Subject> {
    try {
      const result = await this.subjectService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(subjectUpdateSchema))
    body: DeepPartial<subjectType>,
  ): Promise<Subject> {
    console.log(id);
    try {
      return await this.subjectService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Subject> {
    try {
      return await this.subjectService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
