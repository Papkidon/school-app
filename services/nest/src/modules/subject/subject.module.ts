import { Module } from '@nestjs/common';
import { PrismaService } from 'nestjs-prisma';
import { SubjectController } from './subject.controller';
import { SubjectRepository } from './subject.repository';
import { SubjectService } from './subject.service';

@Module({
  controllers: [SubjectController],
  providers: [SubjectService, SubjectRepository, PrismaService],
})
export class SubjectModule {}
