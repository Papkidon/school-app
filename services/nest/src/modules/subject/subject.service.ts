import { Injectable } from '@nestjs/common';
import { SubjectRepository } from './subject.repository';
import { subjectType } from '../../models/subject/subject.types';
import { DeepPartial } from '../../models/utility.types';
import { Subject } from '@prisma/client';
import { subjectsWithClassesDto } from '../../models/subject/subject.types';

@Injectable()
export class SubjectService {
  constructor(private subjectRepository: SubjectRepository) {}
  async findAll(): Promise<Subject[]> {
    return await this.subjectRepository.findAll();
  }

  async findById(id: string): Promise<Subject> {
    return await this.subjectRepository.findById(id);
  }

  async findByName(name: string): Promise<Subject> {
    return await this.subjectRepository.findByName(name);
  }

  async findAllSubjectWithClasses(): Promise<subjectsWithClassesDto> {
    return await this.subjectRepository.findAllSubjectsWithClasses();
  }

  async create(body: any): Promise<Subject> {
    return await this.subjectRepository.create(body);
  }

  async delete(id: string): Promise<Subject> {
    return await this.subjectRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<subjectType>): Promise<Subject> {
    return await this.subjectRepository.update(id, body);
  }
}
