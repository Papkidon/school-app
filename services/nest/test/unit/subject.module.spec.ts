import { Subject } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import {
  mappedSubjectsWithClassesDto,
  subjectsWithClassesDto,
} from '../../src/models/subject/subject.types';
import { SubjectController } from '../../src/modules/subject/subject.controller';
import * as subjectTypes from '../../src/modules/subject/subject.mapper';
import { SubjectRepository } from '../../src/modules/subject/subject.repository';
import { SubjectService } from '../../src/modules/subject/subject.service';
import {
  dummySubjectCreate,
  dummySubjectId,
  dummySubjectUpdate,
} from './dummies/dummy.data';

describe('UserController', () => {
  let subjectController: SubjectController;
  let subjectService: SubjectService;
  let subjectRepository: SubjectRepository;
  let prismaService: PrismaService;

  beforeEach(() => {
    prismaService = new PrismaService();
    subjectRepository = new SubjectRepository(prismaService);
    subjectService = new SubjectService(subjectRepository);
    subjectController = new SubjectController(subjectService);
  });

  afterAll(() => {
    jest.useRealTimers();
    jest.resetAllMocks();
  });

  describe('findAll', () => {
    it('should return an array of all subjects', async () => {
      const result = ['subject1', 'subject2'] as unknown as Subject[];
      jest
        .spyOn(subjectRepository, 'findAll')
        .mockImplementation(async () => result);

      expect(await subjectController.findAll()).toMatchSnapshot();
    });
  });

  describe('findUnique', () => {
    it('should return a subject with given id', async () => {
      const result = 'dummyFoundByIdSubject' as unknown as Subject;
      jest
        .spyOn(subjectRepository, 'findById')
        .mockImplementation(async () => result);

      expect(
        await subjectController.findById(dummySubjectId),
      ).toMatchSnapshot();
    });
  });

  describe('findSubjectsWithClasses', () => {
    it('should return all subjects with their classes', async () => {
      const mappedResult =
        'dummyMappedSubjectsWithClasses' as unknown as mappedSubjectsWithClassesDto;
      const result =
        'dummyFoundSubjectsWithClasses' as unknown as subjectsWithClassesDto;

      jest
        .spyOn(subjectRepository, 'findAllSubjectsWithClasses')
        .mockImplementation(async () => result);

      jest
        .spyOn(subjectTypes, 'mapSubjectsWithClasses')
        .mockImplementation(() => mappedResult);

      expect(
        await subjectController.findAllSubjectsWithClasses(),
      ).toMatchSnapshot();
    });
  });

  describe('create', () => {
    it('should create a new subject', async () => {
      const dummySubjectNotFound = '' as unknown as Subject;
      const result = 'dummyCreatedSubject' as unknown as Subject;
      jest
        .spyOn(subjectRepository, 'create')
        .mockImplementation(async () => result);
      jest
        .spyOn(subjectRepository, 'findByName')
        .mockImplementation(async () => dummySubjectNotFound);

      expect(
        await subjectController.create(dummySubjectCreate),
      ).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('should update a subject with given id', async () => {
      const dummyFoundSubjectById =
        'dummyFoundSubjectById' as unknown as Subject;
      const result = 'dummyUpdatedSubject' as unknown as Subject;
      jest
        .spyOn(subjectRepository, 'update')
        .mockImplementation(async () => result);
      jest
        .spyOn(subjectRepository, 'findById')
        .mockImplementation(async () => dummyFoundSubjectById);

      expect(
        await subjectRepository.update(dummySubjectId, dummySubjectUpdate),
      ).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('should delete a subject with given id', async () => {
      const dummyFoundSubjectById =
        'dummyFoundSubjectById' as unknown as Subject;
      const result = 'dummyDeletedSubject' as unknown as Subject;
      jest
        .spyOn(subjectRepository, 'delete')
        .mockImplementation(async () => result);
      jest
        .spyOn(subjectRepository, 'findById')
        .mockImplementation(async () => dummyFoundSubjectById);

      expect(await subjectController.delete(dummySubjectId)).toMatchSnapshot();
    });
  });
});
