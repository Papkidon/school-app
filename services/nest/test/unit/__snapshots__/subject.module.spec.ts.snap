// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController create should create a new subject 1`] = `"dummyCreatedSubject"`;

exports[`UserController delete should delete a subject with given id 1`] = `"dummyDeletedSubject"`;

exports[`UserController findAll should return an array of all subjects 1`] = `
[
  "subject1",
  "subject2",
]
`;

exports[`UserController findSubjectsWithClasses should return all subjects with their classes 1`] = `"dummyMappedSubjectsWithClasses"`;

exports[`UserController findUnique should return a subject with given id 1`] = `"dummyFoundByIdSubject"`;

exports[`UserController update should update a subject with given id 1`] = `"dummyUpdatedSubject"`;
